compiler/damlc/lib/DA/Cli/Damlc/DependencyDb.hs

```diff
@@ -0,0 +1,251 @@
+-- Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+module DA.Cli.Damlc.DependencyDb
+    ( installDependencies
+    , dependenciesDir
+    , queryDalfsFromDependencies
+    , queryDalfsFromDataDependencies
+    , queryMainDalfs
+    , mainsDir
+    , dalfsDir
+    , configsDir
+    , sourcesDir
+    , normalDepsDir
+    ) where
+
+import qualified "zip-archive" Codec.Archive.Zip as ZipArchive
+import Control.Exception.Safe (tryAny)
+import Control.Monad.Extra
+import DA.Daml.Compiler.Dar
+import DA.Daml.Compiler.DecodeDar (DecodedDalf(..), decodeDalf)
+import DA.Daml.Compiler.ExtractDar (ExtractedDar(..), extractDar)
+import qualified DA.Daml.LF.Ast as LF
+import DA.Daml.Options.Types
+import DA.Daml.Package.Config
+import qualified DA.Pretty
+import Data.Aeson (eitherDecodeFileStrict', encode)
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Lazy as BSL
+import Data.List.Extra
+import qualified Data.Set as Set
+import qualified Data.Text.Extended as T
+import Development.IDE.Types.Location
+import GHC.Fingerprint
+import System.Directory.Extra
+import System.FilePath
+import System.IO.Extra
+
+-- Dependency Database Layout
+-----------------------------
+-- Here is an exemplary dependecy database:
+-- .
+-- ├── data-deps
+-- │   ├── configs
+-- │   │   └── proj2-0.0.1.conf
+-- │   ├── dalfs
+-- │   │   ├── 057eed1fd48c238491b8ea06b9b5bf85a5d4c9275dd3f6183e0e6b01730cc2ba
+-- │   │   │   └── daml-stdlib-DA-Internal-Down-057eed1fd48c238491b8ea06b9b5bf85a5d4c9275dd3f6183e0e6b01730cc2ba.dalf
+-- │   │   ├── 3c1af853a9bc7d6aa214b98f0e4d2099187bd5fa7ab07852b4165b84a82325d3
+-- │   │   │   └── proj2-0.0.1-3c1af853a9bc7d6aa214b98f0e4d2099187bd5fa7ab07852b4165b84a82325d3.dalf
+-- │   │   └── e9eeb8cf890e5a15bc65031f1a8373a4bb7f89229473dabaa595c25d35768b21
+-- │   │       └── daml-prim-e9eeb8cf890e5a15bc65031f1a8373a4bb7f89229473dabaa595c25d35768b21.dalf
+-- │   ├── mains
+-- │   │   └── 3c1af853a9bc7d6aa214b98f0e4d2099187bd5fa7ab07852b4165b84a82325d3
+-- │   │       └── proj2-0.0.1-3c1af853a9bc7d6aa214b98f0e4d2099187bd5fa7ab07852b4165b84a82325d3.dalf
+-- │   ├── sdk-version
+-- │   └── sources
+-- │       └── proj2-0.0.1-3c1af853a9bc7d6aa214b98f0e4d2099187bd5fa7ab07852b4165b84a82325d3
```

Comment:
Same qestion here as for `configs`.

Reply:
that would mean it's not flat anymore and we need to do a traversal when we set up the package db.

Reply:
I guess when I said "flat" I meant that a dalf does not contain its dependencies. Grouping together data for the same dalf (the actual dalf, configs, sources, …) seems fine? Yes we have to do a traversal but we already traverse everything once to find all dalfs, doing another doesn’t seem like an issue (and we could merke them into one).